src/weights.jl

```diff
@@ -569,47 +565,62 @@ function quantile(v::RealVector{V}, w::AbstractWeights{W}, p::RealVector) where
         x < 0 && error("weight vector cannot contain negative entries")
     end
 
-    # full sort
-    vw = sort!(collect(zip(v, w.values)))
+    wvalues = w.values
+    nz = find(w.values)
+    #normalize if non frequencyweight
+    if !isa(w, FrequencyWeights)
+        wvalues = wvalues / w.sum
+    end
+    wsum = sum(wvalues)
 
-    wsum = w.sum
+    #remove zeros weights and sort
+    vw = sort!(collect(zip(view(v, nz), view(wvalues, nz))))
+    N = length(vw)
 
     # prepare percentiles
     ppermute = sortperm(p)
     p = p[ppermute]
     p = bound_quantiles(p)
 
     # prepare out vector
-    N = length(vw)
     out = Vector{typeof(zero(V)/1)}(length(p))
     fill!(out, vw[end][1])
 
     # start looping on quantiles
-    cumulative_weight, Sk, Skold =  zero(W), zero(W), zero(W)
-    vk, vkold = zero(V), zero(V)
-    k = 1
+    Sk, Skold =  zero(W), zero(W)
+    vk, vkold, cumwk, wk = zero(V), zero(V), zero(V), zero(V)
+    k = 0
+
+
     for i in 1:length(p)
-        h = p[i] * (N - 1) * wsum
-        if h == 0
-            # happens when N or p or wsum equal zero
-            out[ppermute[i]] = vw[1][1]
-        else
+        if isa(w, FrequencyWeights)
```

Comment:
It feels weird to use a completely different path for frequency weights. Couldn't a common path be defined, moving some type-specific computations out of the loop like you did for normalization? For example, for `h = p[i] * (wsum - 1) + 1`, `wsum` just needs to be replaced with `N` for non frequency weights, so you could as well define the variable before the loop.

BTW, wouldn't it make more sense to normalize the weights to sum to `N` (rather than 1), since it plays a role equivalent to `wsum`?

Reply:
I have thought about this but I think it is better to do two different paths. Joining the two looks more confusing than enlightening.